src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -74,13 +74,29 @@ def resolve(self, root_reqs, check_supported_wheels):
         requirements = []
         for req in root_reqs:
             if req.constraint:
-                assert req.name
-                assert req.specifier
+                # TODO: Add warnings to accompany these errors, explaining
+                # that these were undocumented behaviour of the old resolver
+                # and will be removed in the new resolver. We need to consider
+                # how we remember to remove these warnings when the new
+                # resolver becomes the default...
+                if not req.name:
+                    raise InstallationError(
+                        "Unnamed requirements are not allowed as constraints"
+                    )
+                if req.link:
+                    raise InstallationError(
+                        "Links are not allowed as constraints"
+                    )
+                if req.extras:
+                    raise InstallationError(
+                        "Constraints cannot have extras"
+                    )
+                specifier = req.specifier or SpecifierSet()
```

Comment:
Can we disallow this as well?

Reply:
We can. There's a test for it, and it's harmless, but yes, we could reject it.

Reply:
It has no effect so... I don't see why we should let this stay when we're tidying up. :)

Reply:
Honestly I favour keeping this in since the impact on implementation is so minimal.

Reply:
On reflection, I'm going to go with @uranusjr on this. Unlike the other cases, this needs no extra logic in the main part of the code, and it's just a degenerate case of the supported specifier syntax, so I don't feel there's any particular reason not to support it.

Reply:
I think `req.specifier` is guarenteed to be a `SpecifierSet` (even for a URL requirement, in which case it would be an empty `SpecifierSet()`, so the `or` part is redundant. If we want to make sure the specifier exist, I’d add an assertion here instead and call it a `packaging` bug if that ever fails.

Reply:
I'd still like to disallow no-qualifiers SpecifierSet. :/

Anyway, that's something we can do in a follow up, so we'll talk about it then.

Reply:
I'm gonna mark this as resolved for the "empty" specifier thing to be tackled later, moved @uranusjr's (correct) suggestion into a separate "thread" on the same line.